### DISCUSSION TITLE
Kafka offset out of sync cause Keda to never scale down our services

### DISCUSSION THREAD

**Discussion by m4nu56** [kedacore/keda#4738] [General] [answered]
# Describe the issue

I don't think this is a "keda issue" but I do think this community will be able to help me solve this problem.

We've got 2 services configured to scale with Keda (v2.6.1, we're upgrading to 2.11 this week) with a Kafka trigger. Config is pretty basic: 
```yaml
spec:
  cooldownPeriod: 300
  maxReplicaCount: 8
  minReplicaCount: 1
  pollingInterval: 30
  scaleTargetRef:
    name: my-service
  triggers:
    - metadata:
        allowIdleConsumers: 'false'
        bootstrapServers: kafka-svc-kafka-external-bootstrap.strimzi.svc:9092
        consumerGroup: my-consumer-group
        lagThreshold: '1'
        offsetResetPolicy: earliest
        scaleToZeroOnInvalidOffset: 'false'
        topic: my-topic
      type: kafka
```

And we can see in the consumer group metadata on the broker that one of the partition doesn't have any offset commited: 

```
kafka-consumer-groups --bootstrap-server kafka-svc-kafka-external-bootstrap.strimzi.svc:9092 --group my-consumer-group --describe

GROUP                                    TOPIC                           PARTITION  CURRENT-OFFSET  LOG-END-OFFSET  LAG             CONSUMER-ID                                                                HOST            CLIENT-ID
my-consumer-group my-topic 5          294             294             0               my-service-kafka-listener-c1da66b6-b202-425a-a121-45f97caf9d8f /10.1.36.27     my-service-kafka-listener
my-consumer-group my-topic 2          -               53              -               my-service-kafka-listener-3e87e280-29bf-4137-b093-035b817793e8 /10.1.36.27     my-service-kafka-listener
my-consumer-group my-topic 6          285             285             0               my-service-kafka-listener-d4c35629-3e74-4f47-aff9-0a3c0ab94974 /10.1.36.27     my-service-kafka-listener
my-consumer-group my-topic 1          516             516             0               my-service-kafka-listener-32b56b37-a434-49ba-9a02-6544409aaba3 /10.1.36.11     my-service-kafka-listener
my-consumer-group my-topic 7          332             332             0               my-service-kafka-listener-ddf73e47-f5ee-45f1-9aea-b25edbb603f3 /10.1.36.27     my-service-kafka-listener
my-consumer-group my-topic 3          102             102             0               my-service-kafka-listener-8dc22a15-0901-44a3-ad01-cccdb305962c /10.1.36.25     my-service-kafka-listener
my-consumer-group my-topic 0          480             480             0               my-service-kafka-listener-0b883fa6-6cb9-477a-a088-5e3adc262ae0 /10.1.36.27     my-service-kafka-listener
my-consumer-group my-topic 4          247             247             0               my-service-kafka-listener-bdb10815-020d-476f-b130-f1f3ec7a7504 /10.1.36.27     my-service-kafka-listener
```

We can see in “debug” logs of keda-operator that the lag for partition 2 is 53 due to the current-offset of this partition for this group being unsync. 

I know for sure that these messages have been properly handled and I have no reason to believe they have not been committed afterwards. 

We are encountering this issue with our 2 services running with Keda and in most of our environments where we deploy these. All in AKS. 

Would you have any clue at where I could start looking at to pinpoint the root cause of this issue ? 

Thanks a lot

**Comment by m4nu56** [comment, marked as answer]
In case anyone ends up in this dark issue, here is how we solved it: 

# Root Cause Analysis

> tldr: when the group is deleted keda will just use the number of messages in each partition to determine the lag. 
> But the group only gets deleted after the offset for the group are lost (ie: after the offsets.retention.minutes is triggered)
> So the only solution we have is to increase this retention period and we should be good !

Kafka, by default deletes committed offsets after a configurable period of time. See parameter `offsets.retention.minutes`. Ie, if a consumer group is inactive (ie, does not commit any offsets) for this amount of time, the offsets get deleted. Thus, even if the consumer is running, if it does not commit offsets for some partitions, those offsets are subject to `offsets.retention.minutes`.

If you start a consumer, the following happens:

* look for a (valid) committed offset (for the consumer group)
* if valid offset is found, resume from there
* if no valid offset is found, reset offset according to auto.offset.reset parameter

Thus, if your offsets got deleted and `auto.offset.reset = latest`, you consumer will not poll anything until new data is added to the topic. If `auto.offset.reset = earliest` it should consume the whole topic.

What is happening in our case is that no messages are committed on those topics for a long period of times, given the environment is not heavily used.

Last run in one of our region was on June 22 and the pods have been up for 5+ days. So 7 retention period + 5 days would do 4th of July. Since June 29th the offsets have been reset due to the retention policy and when Keda calculates the lag between the number of messages in the topic vs the last committed offset. Since this offset is now null/unset the lag equals the number of messages in the topic. 

Current config of the broker can be found using:

```
kafka-configs --bootstrap-server kafka-light.staging.sophiagenetics.internal:9092 --entity-type brokers --entity-name 0 --all --describe
```

we can find: 

```
offsets.retention.minutes=10080 sensitive=false synonyms={DEFAULT_CONFIG:offsets.retention.minutes=10080}
log.retention.hours=48 sensitive=false synonyms={STATIC_BROKER_CONFIG:log.retention.hours=48, DEFAULT_CONFIG:log.retention.hours=168}
```

And no property `auto.offset.reset`

What I suggest is to increase the `offsets.retention.minutes` to `5259600` **(10 years)**
